src/app/config/confparse.c

```diff
@@ -405,28 +440,42 @@ config_validate_single(const config_format_t *fmt,
  * though in general they should only update derived fields rather than fields
  * managed by a configuration format.
  *
- * On success, return 0.  On failure, return -1 and set *<b>msg_out</b> to a
- * description of the problem.
+ * On success, return 0.  On failure, return -1 and add strings to
+ * <b>errs_out</b> describing the problems encountered.  If <b>errs_out</b> is
+ * NULL, messages are simply logged.
  */
-int
+config_validate_status_t
 config_validate(const config_mgr_t *mgr,
-                const void *oldval, void *newval, char **msg_out)
+                const void *oldval, void *newval, smartlist_t *errs_out)
 {
-  if (config_validate_single(mgr->toplevel, oldval, newval, msg_out) < 0) {
-    return -1;
-  }
+  config_validate_status_t res = CFG_OK;
+  smartlist_t *errs_tmp = NULL;
+  if (!errs_out)
+    errs_out = errs_tmp = smartlist_new();
+
+  res = config_validate_single(mgr->toplevel, oldval, newval, errs_out);
+  if (res != CFG_OK)
+    goto done;
 
   SMARTLIST_FOREACH_BEGIN(mgr->subconfigs, const config_format_t *, fmt) {
     const void *old_obj = NULL;
     if (oldval)
       old_obj = config_mgr_get_obj(mgr, oldval, fmt_sl_idx);
     void *new_obj = config_mgr_get_obj_mutable(mgr, newval, fmt_sl_idx);
-    if (config_validate_single(fmt, old_obj, new_obj, msg_out) < 0) {
-      return -1;
-    }
+    res = config_validate_single(fmt, old_obj, new_obj, errs_out);
+    if (res != CFG_OK)
+      goto done;
```

Comment:
Now I have the converse question:

Why do we stop here on failure, rather than collecting all the errors?
We should either do all the validation, or stop on the first error.
Consistency is important here.

Please also document this behaviour in each relevant function's comments, including the callback function's comments.

This change should be made shortly after merging this branch, because there should not be any more branches until after this branch is merged, and the outstanding bugs are fixed.